File: 27-system-ai/5-ai-math-fundamentals/main.cpp
// AI Math Fundamentals - C++ example (Matrix multiplication)

#include <iostream>
#include <vector>

using namespace std;

vector<vector<int>> matMul(const vector<vector<int>>& A, const vector<vector<int>>& B) {
    int n = A.size(), m = B[0].size(), p = B.size();
    vector<vector<int>> C(n, vector<int>(m, 0));

    for(int i=0; i<n; i++) {
        for(int j=0; j<m; j++) {
            for(int k=0; k<p; k++) {
                C[i][j] += A[i][k] * B[k][j];
            }
        }
    }
    return C;
}

int main() {
    vector<vector<int>> A = {{1,2},{3,4}};
    vector<vector<int>> B = {{5,6},{7,8}};
    vector<vector<int>> C = matMul(A, B);

    cout << "Matrix multiplication result:" << endl;
    for(auto& row : C) {
        for(auto val : row) cout << val << " ";
        cout << endl;
    }
    return 0;
}

File: 27-system-ai/5-ai-math-fundamentals/main.java
// AI Math Fundamentals - Java example (Matrix multiplication)

public class Main {
    public static int[][] matMul(int[][] A, int[][] B) {
        int n = A.length;
        int m = B[0].length;
        int p = B.length;
        int[][] C = new int[n][m];

        for(int i = 0; i < n; i++) {
            for(int j = 0; j < m; j++) {
                for(int k = 0; k < p; k++) {
                    C[i][j] += A[i][k] * B[k][j];
                }
            }
        }
        return C;
    }

    public static void main(String[] args) {
        int[][] A = {{1, 2}, {3, 4}};
        int[][] B = {{5, 6}, {7, 8}};
        int[][] C = matMul(A, B);

        System.out.println("Matrix multiplication result:");
        for (int[] row : C) {
            for (int val : row) {
                System.out.print(val + " ");
            }
            System.out.println();
        }
    }
}
